### PR TITLE
[WIP] Focused tests

### PR DESCRIPTION
**Pull request by brianium**
This is a work in progress, to implement #185

### CODE REVIEW

File: src/Dsl.php [hunk 1]
```diff
@@ -20,7 +20,7 @@ function describe($description, callable $fn)
  */
 function context($description, callable $fn)
 {
-    describe($description, $fn);
+    Context::getInstance()->addSuite($description, $fn);
```
Comment (brianium) [inline src/Dsl.php]: Are we just favoring explicitness here? I always kind of viewed context and all of its forms as just an alias for describe. Why not just have it delegate to the describe function?

File: src/Core/Suite.php [hunk 2]
```diff
@@ -111,4 +127,23 @@ protected function runTest(TestInterface $test, TestResult $result)
         $test->setEventEmitter($this->eventEmitter);
         $test->run($result);
     }
+
+    private function focusedTests()
```
Comment (brianium) [inline src/Core/Suite.php]: Couple thoughts on this:

I don't think I ever published any real coding standards, so I am open to being shot down on this :) - but for this project (and generally in the PHP realm for me) - I tended more towards the `get` language for things returning something. It is a little more verbose, but I think it fits the language and the rest of the project a little better. Something like `getFocusedTests`. In JavaScript and other functional languages I use, I like to ditch some of that verbosity. Thoughts?

Second, I'm not sure highlighting `focused` here is the best solution. It is currently our only need so again I'm probably wrong :). It kind of seems like we are applying a transformation of state to the test tree before we run things, and this seems like something that might fit well into an `array_filter` type scenario:

```php
// or maybe getTestsToRun?
protected function getFilteredTests()
{
    $tests = array_filter($this->tests, function (TestInterface $test) {
        return $test->getFocused();
    });

    return empty($tests) ? $this->tests : $tests;
}
```

Basically, if the state of the things to run has changed we know we can return that changed set. Otherwise return the original. The language might be a bit more flexible if we add similar functionality later on, and it also eliminates the need for the additional `$hasFocusedTests` variable. Thoughts?

File: specs/suite.spec.php [hunk 3]
```diff
@@ -197,6 +221,46 @@
...
+    describe('->getFocused()', function() {
+        context('when explicitly marked as focused', function () {
+            beforeEach(function() {
+                $this->suite = new Suite("test suite", function() {}, true);
+            });
+
+            it('should return true even if nested tests are not focused', function() {
+                $test = new Test("test", function() {});
+                $this->suite->addTest($test);
+                assert($this->suite->getFocused(), "suite should not be focused");
```
Comment (ezzatron) [inline specs/suite.spec.php]: Yep, good catch.

File: src/Core/Suite.php (hunk 2 above)
Comment (ezzatron) [inline src/Core/Suite.php]: Yeah, I have no problem with any of that. I tend not to use the "get" prefix in my own stuff, but I know Peridot does tend to use it, so I think I just did it out of habit.

Having a more generic name for the method makes total sense, as does the filter approach. It also becomes more relevant if we change that method from `private` to `protected`, as it could then be interpreted as being part of Peridot's API, so having a future-proof name is a good idea.

File: src/Core/TestInterface.php [hunk 4]
```diff
@@ -90,6 +90,13 @@ public function getPending();
     public function setPending($state);
 
     /**
+     * Return whether or not the test is focused
+     *
+     * @return bool
+     */
+    public function getFocused();
```
Comment (ezzatron) [inline src/Core/TestInterface.php]: Definitely, but that would also be inconsistent with `getPending()`. In light of that, what do you think I should call it?

File: src/Dsl.php (hunk 1 above)
Comment (ezzatron) [inline src/Dsl.php]: Yeah, I probably shouldn't have changed that here. My logic was "less method calls = more faster", especially for PHP versions < 7. In any case, it belongs in its own PR.

**Comment by ezzatron** [comment]
Thanks for the review man, I'll probably be able to get your suggestions in today sometime.

File: src/Core/TestInterface.php (hunk 4 above)
Comment (brianium) [inline src/Core/TestInterface.php]: I remember there being some reasoning to `getPending()` . I think I remember it having something to do with pending being one of three states instead of just true/false.

I think the original thought was that null means a pending state does not exist, where true/false are explicit pending states. Hence `Suite.php` has the following:

```php
    protected function runTest(TestInterface $test, TestResult $result)
    {
        if ($this->getPending() !== null) {
            $test->setPending($this->getPending());
        }

        $test->setEventEmitter($this->eventEmitter);
        $test->run($result);
    }
```

Looking at it now, I can't remember why the explicit setting was favored....

Comment (ezzatron) [inline src/Core/TestInterface.php]: Yeah, that makes some degree of sense. Since the focused state doesn't use `null`, it can probably be `isFocused()` then. I'll update that in a sec.

**Comment by brianium** [comment]
@ezzatron what is your stance on squashing commits? I tend to lean more towards cleaning up commit messages addressing review feedback, but I am open to changing that mindset if you have a preference

Thanks again man!

**Comment by ezzatron** [comment]
I tend to squash once all feedback has been addressed. I think I covered all your review points, so I can do that now if that works for you?
